# Assignment 2 Part A
Name: David Fodor
Student code: B00796884
Email: fodor-d@ulster.ac.uk

## Preprocessing
## Identifying target and features

In [18]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Importing the dataset
dataset = pd.read_csv('diagnose5.csv')
#Specifying the features X and the target y
X = dataset.iloc[:, 3:16].values
y = dataset.iloc[:, 16].values
#Printing the dataset
print(dataset)
display(dataset.describe())

      ID  Gender Location           x1           x2           x3           x4  \
0      0    Male   Dublin  -467.074369  -184.726516  -967.684803   928.017233   
1      1  Female   Dublin   232.526054  -524.912660   895.377400   335.754492   
2      2  Female   London   962.877057   -45.760582          NaN   172.804295   
3      3    Male   London  1274.585571 -1184.654089   170.258993   753.411953   
4      4  Female  Belfast   875.147413  -169.150411   464.195000 -1381.642828   
..   ...     ...      ...          ...          ...          ...          ...   
347  347  Female   London   -61.519302  1438.553254  1582.568494  -666.954544   
348  348  Female   London   139.640327 -1558.417473   -31.728934  1265.082189   
349  349    Male   Dublin  -510.964488  -980.392637  -384.941829   667.428958   
350  350  Female  Belfast   -18.396341  1329.757179  1093.447103  -323.641412   
351  351    Male  Belfast  -970.188522 -1553.758612  -104.805284   196.881533   

              x5           

ID           x1           x2           x3           x4  \
count  352.000000   351.000000   352.000000   351.000000   352.000000   
mean   175.500000   -77.311644   310.492082    18.181941    48.596349   
std    101.757883   847.965600  1392.940015   905.530758   957.229191   
min      0.000000 -3050.818857 -4558.753586 -2452.149474 -2372.677643   
25%     87.750000  -639.727179  -636.880928  -608.147095  -588.971097   
50%    175.500000   -98.234844   348.884873    67.024286    38.191051   
75%    263.250000   470.140709  1301.838996   658.112650   698.631330   
max    351.000000  2365.867028  3939.210461  2453.807595  3286.966138   

                x5           x6           x7           x8          x9  \
count   351.000000   350.000000   352.000000   351.000000  352.000000   
mean     -5.167915   -25.842429    15.294052   133.363138    0.936987   
std     895.246034   964.472910   402.955200   860.930049   18.039882   
min   -2704.774136 -3332.418575 -1253.844415 -2539.257994  -46.361670   
25%    -637.326133  -628.441026  -264.580232  -455.843682  -11.647162   
50%     -35.053724   -35.403644    10.532292   188.143624    0.700199   
75%     530.149903   630.877068   332.823238   715.870826   13.538374   
max    2576.343059  2780.303151  1130.563491  2406.186955   49.730222   

               x10          x11          x12          x13   Diagnosis  
count   351.000000   352.000000   352.000000   351.000000  352.000000  
mean   -235.813463   -53.136279  -238.915240    22.931604    0.994318  
std    1778.277304   685.181934  1398.032067   440.333567    0.819959  
min   -6492.044080 -2107.070516 -4569.457640 -1208.624182    0.000000  
25%   -1271.796745  -519.786862 -1254.184997  -275.271380    0.000000  
50%    -158.704025   -22.045481  -289.188899   -14.148348    1.000000  
75%     904.455704   354.464640   538.206617   301.796761    2.000000  
max    4704.644731  2226.136291  3634.805699  1273.500813    2.000000

## Missing values

We need to substitute something for the missing values. For this we use the SimpleImputer class from Scikit learn.

In [19]:
from sklearn.impute import SimpleImputer
#Creating a SimpleImputer object, specifying how missing values are represented and our chosen strategy for filling them up
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
#Fitting to the data. Here we specify the relevant features from the matrix X. 
X[:,0:1] = imputer.fit_transform(X[:,0:1])
X[:,2:3] = imputer.fit_transform(X[:,2:3])
X[:,4:6] = imputer.fit_transform(X[:,4:6])
X[:,7:8] = imputer.fit_transform(X[:,7:8])
X[:,9:10] = imputer.fit_transform(X[:,9:10])
X[:,12:13] = imputer.fit_transform(X[:,12:13])
#Printing X to check if it works as it should
print(X)

[[ -467.0743686   -184.7265156   -967.6848028  ...     4.44065212
  -1062.242048    -234.0280502 ]
 [  232.5260537   -524.9126604    895.3774003  ...   228.6765046
    741.7012044    222.117581  ]
 [  962.877057     -45.76058249    18.1819414  ...   -22.35653165
  -1207.530122    -174.2473578 ]
 ...
 [ -510.9644882   -980.3926374   -384.941829   ...  -285.8472407
   1477.733558     438.6458203 ]
 [  -18.39634102  1329.757179    1093.447103   ...   520.3293924
  -1272.396845    -280.2984742 ]
 [ -970.1885224  -1553.758612    -104.8052844  ...   310.5336951
    800.6789307   -411.4117824 ]]


## Feature Scaling

Some features can have very different magnitudes and that can affect the results from machine learning algorithms. The following code scales all of the features.

In [20]:
#Feature Scaling
from sklearn.preprocessing import StandardScaler
#Creating a StandardScaler object
sc = StandardScaler()
#Scaling all of the features
X[:,0:13] = sc.fit_transform(X[:,0:13])
#Printing the results after scaling
print(X)

[[-0.4609559  -0.35602648 -1.0918232  ...  0.08415121 -0.58975672
  -0.58522175]
 [ 0.36643195 -0.60059581  0.97147242 ...  0.41188176  0.70242476
   0.45364306]
 [ 1.23018721 -0.25611999  0.         ...  0.04498595 -0.69382793
  -0.44907212]
 ...
 [-0.51286289 -0.92805309 -0.44644967 ... -0.34011729  1.22965164
   0.94678287]
 [ 0.06967664  0.73277815  1.19082976 ...  0.83814526 -0.74029257
  -0.69060194]
 [-1.05596774 -1.3402619  -0.13620533 ...  0.53151964  0.74467106
  -0.98921053]]


It is a good idea to check the mean of each of the features after the scaling has been done. It should be 0.

In [21]:
print('Mean of x1: {:5.3f}\n'.format(np.mean(X[:,0])))
print('Mean of x2: {:5.3f}\n'.format(np.mean(X[:,1])))
print('Mean of x3: {:5.3f}\n'.format(np.mean(X[:,2])))
print('Mean of x4: {:5.3f}\n'.format(np.mean(X[:,3])))
print('Mean of x5: {:5.3f}\n'.format(np.mean(X[:,4])))
print('Mean of x6: {:5.3f}\n'.format(np.mean(X[:,5])))
print('Mean of x7: {:5.3f}\n'.format(np.mean(X[:,6])))
print('Mean of x8: {:5.3f}\n'.format(np.mean(X[:,7])))
print('Mean of x9: {:5.3f}\n'.format(np.mean(X[:,8])))
print('Mean of x10: {:5.3f}\n'.format(np.mean(X[:,9])))
print('Mean of x11: {:5.3f}\n'.format(np.mean(X[:,10])))
print('Mean of x12: {:5.3f}\n'.format(np.mean(X[:,11])))
print('Mean of x13: {:5.3f}\n'.format(np.mean(X[:,12])))

Mean of x1: 0.000

Mean of x2: -0.000

Mean of x3: -0.000

Mean of x4: 0.000

Mean of x5: -0.000

Mean of x6: 0.000

Mean of x7: 0.000

Mean of x8: -0.000

Mean of x9: -0.000

Mean of x10: 0.000

Mean of x11: 0.000

Mean of x12: 0.000

Mean of x13: -0.000



## Feature Selection

At this point we can select the k best features. After this, we have to modify the features so that they only consist of these selected features.

In [22]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import mutual_info_regression

#Identifying the relevant features
#Through trial and error it was found that for the best accuracy, the value of k should either be 9 or 12
#These two produced the exact same results and since selecting 9 features instead of 12 is less demanding, it is the best solution
select = SelectKBest(mutual_info_regression, k=9).fit(X, y)
#Now transforming the features
X = select.transform(X)
#Printing the features
print(X)

[[-0.4609559  -0.35602648 -1.96025464 ...  0.08415121 -0.58975672
  -0.58522175]
 [ 0.36643195 -0.60059581 -0.03450222 ...  0.41188176  0.70242476
   0.45364306]
 [ 1.23018721 -0.25611999 -0.12221128 ...  0.04498595 -0.69382793
  -0.44907212]
 ...
 [-0.51286289 -0.92805309  0.2658213  ... -0.34011729  1.22965164
   0.94678287]
 [ 0.06967664  0.73277815 -0.41862297 ...  0.83814526 -0.74029257
  -0.69060194]
 [-1.05596774 -1.3402619  -0.7867667  ...  0.53151964  0.74467106
  -0.98921053]]


## Splitting

We split the data into training and test sets using the function train_test_split from the sklearn.model_selection module in Sci-kit learn (sklearn).

In [23]:
from sklearn.model_selection import train_test_split
#The test size is 20% of all the data, and it is selected at random
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)
#Printing out the test data
print(X_test)
print(y_test)

[[ 7.12663946e-01 -1.41863267e+00 -9.94935943e-19  0.00000000e+00
   2.70968169e-01  3.11514523e-01 -1.18150611e+00 -2.10415516e-01
  -2.99506035e-01]
 [-1.43380205e+00 -7.10685099e-02  9.00454064e-01  1.54234773e+00
  -2.65467106e+00  1.02702453e+00 -3.52544909e-01 -1.37662333e+00
   2.32570982e+00]
 [ 5.42891844e-01 -6.38827370e-01  1.57394346e+00 -6.50274688e-02
  -1.48503212e+00 -1.68224259e+00  8.90314589e-01 -1.72032712e-01
  -5.01098535e-01]
 [ 2.25252314e-01  3.74045901e-01  7.55758439e-01 -6.24614299e-01
  -9.38220919e-02  9.38729538e-01 -1.20751053e-01  3.16202148e-02
  -7.12272465e-02]
 [-3.57025318e-01  6.42058458e-01  1.20369992e+00 -1.04822714e+00
  -5.32717323e-01  1.69127228e+00  1.21256467e+00 -6.24825224e-01
   1.69218615e+00]
 [ 9.75449240e-01  1.34247533e+00  1.43486558e+00  1.27765086e+00
  -1.55387289e-01  4.08129112e-01 -2.81340986e-01 -1.87747420e+00
   3.49088919e-01]
 [ 1.29241803e+00 -1.18928131e+00 -1.07610012e+00  5.42947342e-02
   1.13048850e+00  3.2110202

## Training
## Logistic Regression

In [24]:
#Learning Logistic Regression Classifier
from sklearn.linear_model import LogisticRegression
#Creating a LogisticRegression object
lr = LogisticRegression(random_state = 0)
#Fitting the model to the training data
lr.fit(X_train, y_train)
#Predicting test cases
y_pred = lr.predict(X_test)

## Testing and evaluation
## Confusion Matrix

In [25]:
#Constructing the Confusion Matrix
from sklearn.metrics import confusion_matrix
#Creating a confusion_matrix object
cm = confusion_matrix(y_test, y_pred)
#Printing the confusion matrix
print(cm)
#Calculate the accuracy and see if we can get a better result by changing the value of k at the 'Feature selection' step
#Since this is not a binary classification, finding the TP, TN, FP and FN values was a little more difficult.
#A formula was used from the following site:
#https://towardsdatascience.com/confusion-matrix-for-your-multi-class-machine-learning-model-ff9aa3bf7826
#By calculating the accuracy for each of the three cases, and taking the mean of them, we can simplify the formula to this:
accuracy = (3 * cm.item(0) + cm.item(1) + cm.item(2) + cm.item(3) + 3 * cm.item(4) + cm.item(5) + cm.item(6) + cm.item(7) + 3 * cm.item(8) ) / (3 * (cm.item(0) + cm.item(1) + cm.item(2) + cm.item(3) + cm.item(4) + cm.item(5) + cm.item(6) + cm.item(7) + cm.item(8) ) )
#Printing the accuracy
print('Accuracy: ')
print(accuracy)

[[17  8  0]
 [ 0 23  2]
 [ 1  2 18]]
Accuracy: 
0.8779342723004695


## Training
## Decision Tree

In [26]:
#Learning Decision Tree Classifier
from sklearn.tree import DecisionTreeClassifier
#Creating a DecisionTreeClassifier object
dt = DecisionTreeClassifier(criterion = 'entropy', random_state = 0, ccp_alpha = 0.0)
#Fitting the model to the training data
dt.fit(X_train, y_train)
#Predicting test cases
y_pred = dt.predict(X_test)

## Testing and evaluation
## Confusion Matrix

In [27]:
# Constructing the Confusion Matrix
from sklearn.metrics import confusion_matrix
#Fitting the model to the training data
cm = confusion_matrix(y_test, y_pred)
#Printing the confusion matrix
print(cm)
accuracy = (cm.item(0) + cm.item(0) + cm.item(0) + cm.item(1) + cm.item(2) + cm.item(3) + cm.item(4) + cm.item(4) + cm.item(4) + cm.item(5) + cm.item(6) + cm.item(7) + cm.item(8) + cm.item(8) + cm.item(8) ) / (3 * (cm.item(0) + cm.item(1) + cm.item(2) + cm.item(3) + cm.item(4) + cm.item(5) + cm.item(6) + cm.item(7) + cm.item(8) ) )
#Printing the accuracy
print('Accuracy: ')
print(accuracy)

[[17  5  3]
 [ 3 20  2]
 [ 6  0 15]]
Accuracy: 
0.8215962441314554


## Conclusion

If we take a look at the accuracy of both the Logistic Regression and the Decision Tree algorithm, we can compare them and conclude which one produces a more accurate result in case of this dataset. This, of course, is not constant, for as we change the value of k, one might be more accurate than the other and vice versa. By trying out every possible value of k, I have concluded that LR produced a higher accuracy for almost any value of k, except for 4 and 5. In case of k = 4, the accuracy of the two classifiers is identical (0.77). In case of k = 5, the accuracy of DT is a bit higher. However, this hardly matters, for I have concluded earlier that the highest possible accuracy that we can achieve in this scenario (for both LR and DT) is if k has the value of 9. In this case, LR has an accuracy 0f 0.88, while DT has only 0.82. The biggest difference in the accuracy of LR and DT is at k = 10. In this case, LR has 0.88, while DT has only 0.79. In conclusion, LR is the better choice for this scenario. 